fairlearn/adversarial/_adversarial_mitigation.py

```diff
@@ -0,0 +1,822 @@
+        epochs=1,
+        batch_size=-1,
+        max_iter=-1,
+        shuffle=False,
+        progress_updates=None,
+        skip_validation=False,
+        callbacks=None,
+        cuda=None,
+        warm_start=False,
+        random_state=None,
+    ):
+        """Initialize class by only storing (kw)args, as per sklearn API."""
+        self.backend = backend
+        self.predictor_model = predictor_model
+        self.adversary_model = adversary_model
+        self.predictor_loss = predictor_loss
+        self.adversary_loss = adversary_loss
+        self.predictor_function = predictor_function
+        self.predictor_optimizer = predictor_optimizer
+        self.adversary_optimizer = adversary_optimizer
+        self.constraints = constraints
+        self.y_transform = y_transform
+        self.z_transform = z_transform
+        self.learning_rate = learning_rate
+        self.alpha = alpha
+        self.epochs = epochs
+        self.batch_size = batch_size
+        self.max_iter = max_iter
+        self.shuffle = shuffle
+        self.progress_updates = progress_updates
+        self.skip_validation = skip_validation
+        self.callbacks = callbacks
+        self.cuda = cuda
+        self.warm_start = warm_start
+        self.random_state = random_state
+
+    def __setup(self, X, Y, Z):
+        """
+        Initialize the entire model from the parameters and the given data.
+
+        Following sklearn API, we do not do intialization in `__init__`, but
+        instead in `fit`. Firstly, we validate the backend. Then, we validate
+        some key-word arguments. Then, we initialize the BackendEngine, which
+        handles the initialization of the losses and optimizers. Among these
+        steps, if a loss or function is not explicitely defined, we try to
+        infer something appropriate from data (see interpret_keyword).
+        """
+        self._validate_backend()
+
+        # Verify the constraints and set up the corresponding network structure.
+        if self.constraints == "demographic_parity":
+            self.pass_y_ = False
+        elif self.constraints == "equalized_odds":
+            self.pass_y_ = True
+        else:
+            raise ValueError(
+                _KWARG_ERROR_MESSAGE.format(
+                    "constraints", "'demographic_parity' or 'equalized_odds'"
+                )
+            )
+
+        # Non-negative parameters
+        for kw, kwname in (
+            (self.learning_rate, "learning_rate"),
+            (self.alpha, "alpha"),
+            (self.progress_updates, "progress_updates"),
+        ):
+            if kw and kw < 0.0:
+                raise ValueError(
+                    _KWARG_ERROR_MESSAGE.format(kwname, "a non-negative number")
+                )
+
+        # Positive or -1 parameters
+        for kw, kwname in (
+            (self.batch_size, "batch_size"),
+            (self.epochs, "epochs"),
+            (self.max_iter, "max_iter"),
+        ):
```

Comment:
you might wanna check `check_scalar` from sklearn (https://scikit-learn.org/dev/modules/generated/sklearn.utils.check_scalar.html?highlight=check_scalar#sklearn.utils.check_scalar)

Reply:
That's an awfully scary function for such a simple task :)

Reply:
but it makes your code shorter and error messages more consistent.

Reply:
ah okey that makes sense!